src/main/scala/io/iohk/ethereum/forkid/ForkIdValidator.scala

```diff
@@ -0,0 +1,129 @@
+
+    // find the first unpassed fork and it's index
+    val (unpassedFork, i) = (forks :+ maxUInt64).zipWithIndex.find { case (fork, _) => currentHeight < fork }.get
+
+    // The checks are left biased -> whenever a result is found we need to short circuit
+    val validate: F[Option[ForkIdValidationResult]] = (for {
+      _ <- liftF(Logger[F].trace(s"Before checkMatchingHashes"))
+      matching <- fromEither[F](
+        checkMatchingHashes(checksums, remoteId, currentHeight, i).toLeft("hashes didn't match")
+      )
+      _ <- liftF(Logger[F].trace(s"checkMatchingHashes result: $matching"))
+      _ <- liftF(Logger[F].trace(s"Before checkSubset"))
+      sub <- fromEither[F](checkSubset(checksums, forks, remoteId, i).toLeft("not in subset"))
+      _ <- liftF(Logger[F].trace(s"checkSubset result: $sub"))
+      _ <- liftF(Logger[F].trace(s"Before checkSuperset"))
+      sup <- fromEither[F](checkSuperset(checksums, remoteId, i).toLeft("not in superset"))
+      _ <- liftF(Logger[F].trace(s"checkSuperset result: $sup"))
+      _ <- liftF(Logger[F].trace(s"No check succeeded"))
+      res <- fromEither[F](Either.left[ForkIdValidationResult, Unit](ErrLocalIncompatibleOrStale))
+    } yield (res)).value
+      .map(_.swap)
+      .flatMap(res => Logger[F].debug(s"Validation result is: $res") >> Monad[F].pure(res.toOption))
+
+    Logger[F].debug(s"Validating $remoteId") >>
+      Logger[F].trace(s"Forks list: $forks") >>
+      Logger[F].trace(s"Unpassed fork $unpassedFork was found at index $i") >>
+      validate.map(_.getOrElse(Connect)) // Impossible to say if nodes are compatible, so we need to allow to connect
+  }
+
+  private def calculateCheckchecksums(
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  ): Vector[BigInt] = {
+    val crc = new CRC32()
+    crc.update(genesisHash.asByteBuffer)
+    val genesisChecksum = BigInt(crc.getValue())
+
+    genesisChecksum +: (forks.map { fork =>
+      crc.update(bigIntToBytes(fork, 8))
+      BigInt(crc.getValue())
+    }).toVector
+  }
+
+  /**
+    * 1) If local and remote FORK_HASH matches, compare local head to FORK_NEXT.
+    * The two nodes are in the same fork state currently.
+    * They might know of differing future forks, but that’s not relevant until the fork triggers (might be postponed, nodes might be updated to match).
+    * 1a) A remotely announced but remotely not passed block is already passed locally, disconnect, since the chains are incompatible.
+    * 1b) No remotely announced fork; or not yet passed locally, connect.
+    */
+  private def checkMatchingHashes(
+      checksums: Vector[BigInt],
+      remoteId: ForkId,
+      currentHeight: BigInt,
+      i: Int
+  ): Option[ForkIdValidationResult] = {
+    if (checksums(i) == remoteId.hash) {
+      if (remoteId.next.isDefined && currentHeight >= remoteId.next.get) {
+        Some(ErrLocalIncompatibleOrStale)
+      } else {
+        Some(Connect)
+      }
+    } else { None }
+  }
```

Comment:
Could be nicer as a match on `remoteId`:

```scala
remoteId match {
  case ForkId(hash, _) if checksums(i) != hash => None
  case ForkId(_, Some(next)) if currentHeight >= next => Some(ErrLocalIncompatibleOrStale)
  case _ => Some(Connect)
}
```

(Gets rid of the nested `if` and the `Option.get`.)